sci-log-db/src/mixins/basesnippet.repository-mixin.ts

```diff
@@ -0,0 +1,378 @@
+
+function UpdateAndDeleteRepositoryMixin<
+  M extends Entity & {
+    expiresAt: Date;
+    versionable?: boolean;
+    parentId?: string;
+    id: string;
+    deleted?: boolean;
+    snippetType: string;
+  },
+  ID,
+  Relations extends object,
+  R extends MixinTarget<DefaultCrudRepository<M, ID, Relations>>
+>(superClass: R) {
+  class Mixed extends superClass {
+    @repository.getter('BasesnippetRepository')
+    readonly baseSnippetRepositoryGetter: Getter<BasesnippetRepository>;
+
+    private async baseSnippetRepository(): Promise<BasesnippetRepository> {
+      const basesnippetRepository =
+        this.constructor.name === 'BasesnippetRepository'
+          ? this
+          : await this.baseSnippetRepositoryGetter();
+      return basesnippetRepository as BasesnippetRepository;
+    }
+
+    async updateByIdWithHistory(
+      id: ID,
+      basesnippet: Basesnippet,
+      user: UserProfile,
+    ): Promise<void> {
+      if (
+        typeof basesnippet.deleted == 'undefined' ||
+        basesnippet.deleted === false
+      ) {
+        const snippet = await this.findById(id, {}, {currentUser: user});
+        if (
+          (typeof snippet?.expiresAt != 'undefined' &&
+            snippet.expiresAt.getTime() < Date.now()) ||
+          typeof snippet?.expiresAt == 'undefined'
+        ) {
+          throw new HttpErrors.Forbidden('Cannot modify expired data snippet.');
+        }
+        if (snippet?.versionable) {
+          await this.addToHistory(snippet, user);
+        }
+      }
+      await this.updateById(id, basesnippet, {
+        currentUser: user,
+      });
+    }
+
+    private async addToHistory(snippet: M, user: UserProfile): Promise<void> {
+      const historySnippet = await this.getHistorySnippet(snippet, user);
+      const snippetCopy = _.omit(snippet, 'id');
+      snippetCopy.parentId = historySnippet.id;
+      const baseSnippetRepository = await this.baseSnippetRepository();
+      await baseSnippetRepository.create(snippetCopy, {currentUser: user});
+    }
+
+    private async getHistorySnippet(
+      snippet: M,
+      user: UserProfile,
+    ): Promise<Basesnippet> {
+      const baseSnippetRepository = await this.baseSnippetRepository();
+      let historySnippet = await baseSnippetRepository.findOne(
+        {where: {snippetType: 'history', parentId: snippet.id}},
+        {currentUser: user},
+      );
+
+      if (historySnippet == null) {
+        const historySnippetPayload = _.pick(snippet, [
+          'isPrivate',
+          'ownerGroup',
+          'accessGroups',
+        ]);
+        historySnippetPayload.parentId = snippet.id;
+        historySnippetPayload.snippetType = 'history';
+        historySnippet = await baseSnippetRepository.create(
+          historySnippetPayload,
+          {currentUser: user},
+        );
+      }
+      return historySnippet;
+    }
+
+    async deleteByIdWithHistory(id: ID, user: UserProfile): Promise<void> {
```

Comment:
I couldn't really understand the logic of this method. For me, it would have made more sense to apply the same logic we apply on update

Reply:
The problem with delete endpoints is that mongo's changestream does not broadcast the information properly, hence the two-step process of announcing the deletion and the actual deletion.